pkg/volume/csi/csi_util.go

```diff
@@ -140,3 +137,57 @@ func hasReadWriteOnce(modes []api.PersistentVolumeAccessMode) bool {
 	}
 	return false
 }
+
+// getSourceFromSpec returns either CSIVolumeSource or CSIPersistentVolumeSource, but not both
+func getSourceFromSpec(spec *volume.Spec) (*api.CSIVolumeSource, *api.CSIPersistentVolumeSource, error) {
+	if spec == nil {
+		return nil, nil, fmt.Errorf("volume.Spec nil")
+	}
+	if spec.Volume != nil && spec.Volume.CSI != nil && utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+		return spec.Volume.CSI, nil, nil
+	}
+	if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.CSI != nil {
+		return nil, spec.PersistentVolume.Spec.CSI, nil
+	}
+
+	return nil, nil, fmt.Errorf("volume source not found in volume.Spec")
+}
+
+// getPVSourceFromSpec ensures only CSIPersistentVolumeSource is present in volume.Spec
+func getPVSourceFromSpec(spec *volume.Spec) (*api.CSIPersistentVolumeSource, error) {
+	volSrc, pvSrc, err := getSourceFromSpec(spec)
+	if err != nil {
+		return nil, err
+	}
+	if volSrc != nil {
+		return nil, fmt.Errorf("unexpected api.CSIVolumeSource found in volume.Spec")
+	}
+	return pvSrc, nil
+}
+
+// getCSIFieldsFromSpec returns CSI fields either from CSIVolumeSource or CSIPersistentVolumeSource
+func getCSIFieldsFromSpec(spec *volume.Spec) (driver, volumeHandle, fsType string, readOnly bool, err error) {
```

Comment:
it looks like this is only used for tests? if so, can we put it in the test file instead?